### PR TITLE
Fix CBC padding check.

### PR DESCRIPTION
**Pull request by boaks**
Check all padding bytes for the right values.
 
Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by boaks** [comment]
New approach:
Decouple used compare and MAC calculations from padding length.
Always call padding compare twice, with 256 compares in sum.
Also calculate extra MAC blocks to have a fixed sum of blocks independent from the padding length.

**Comment by sophokles73** [comment]
@boaks if you want me to review this, I'll need an introduction to the ideas behind it first :-)

**Comment by boaks** [comment]
[page 16-17](http://www.isg.rhul.ac.uk/tls/TLStiming.pdf)

But that specification seem to be based on 13 bytes application data (step 5 on page 17), which is not valid for TLS12_CID (which includes the cid in the additional data). And using always the 256 bytes for "padding comparison" even if the "plain text/plen" is smaller than 256 (step 3 on page 16), seems for me currently also not be the right choice.

If "very similar processing times" should be achieved with that, I think we need also some test to verify that. But I have doubts, that a shared "jenkins" will be able to run such tests. With that, I'm got aware, that it will take more time to complete this work. 
And so for now, I switched back to other tasks (as PR #1002) or my "scandium clean up work" (issue #905),

**Comment by boaks** [comment]
I personally prefer, to just merge the 1. commit and leave the other changes unmerged.
The CBC attacks assumes, that a series of records is also processed as series, and the last record is answered by a response.
With scandium 2.0 the records of different associations are processed in parallel, so the series will not be kept.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/InvalidMacException.java [hunk 1]
```diff
@@ -18,25 +18,46 @@
...
+
+	/**
+	 * Create invalid MAC exception without actual MAC values.
+	 * 
+	 * @param msg message
+	 */
+	public InvalidMacException(String msg) {
+		super(msg);
+		this.expected = Bytes.EMPTY;
+		this.actual = Bytes.EMPTY;
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/InvalidMacException.java]: You don't use this constructor is it intended ? (maybe you planned to add a more specific message for padding error above ?)

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java [hunk 2]
```diff
@@ -596,15 +598,22 @@ private Record(ProtocolVersion version, int epoch, long sequenceNumber, long rec
 				new IvParameterSpec(iv));
 		byte[] plaintext = blockCipher.doFinal(reader.readBytesLeft());
 		// last byte contains padding length
-		int paddingLength = plaintext[plaintext.length - 1];
+		int paddingLength = plaintext[plaintext.length - 1] & 0xff;
 		int fragmentLength = plaintext.length
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java]: Just a detail but this is no more the fragment length ? (but I can find a better name)

Maybe we can keep :
```java
int fragmentLength = plaintext.length
				- 1 // paddingLength byte
				- paddingLength
				- currentReadState.getMacLength();
and check if 
``` java
if (fragmentLength < 0) {
	throw new InvalidMacException();
}
```

**Comment by boaks** [comment]
That's from the description of the attack:

"In DTLS, there are no error messages, but the techniques of [1] can be applied to solve this problem. There, the authors send a packet containing a ciphertext C closely followed by a DTLS message, with the latter always provoking a response message."

So, for at least the "multi-threaded" scandium 2.0, I can't see, that this attack makes sense, because the series of records is processed in parallel by multiple threads.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java (hunk 2 above)
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java]: If we agree to go for the  1 commit only, yes.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/InvalidMacException.java (hunk 1 above)
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/InvalidMacException.java]: It's used by

```
public InvalidMacException() {
	this(DEFAULT_MESSAGE);
}
```

(even if that may be replaced.)

**Comment by sbernard31** [comment]
My general feeling : I understand why you are perplex about the timing attack (in multithread/cluster environement)  but I'm not so confident about my skills in security, so it's hard to me to bypass security expert recommendation.

Some concerns : 
1. scandium 2.0 is multi-threaded but as all traffic from the same peer is executed sequentially, so maybe it is possible to use the tricks described above ?
2. in cluster environment, I don't know if everybody to this, but as scandium does not have shared state for DTLS connection, we generally try to stick 1 client with 1 cluster instance. So the cluster in this case will not bring so much variability ?

**Comment by sbernard31** [comment]
(@ic0ns, @jurajsomorovsky just in case, you are interested by this :slightly_smiling_face:)

**Comment by boaks** [comment]
"same peer"
So your "on-path-attacker" sees a sequence of records, manipulates some and some not, and waits for the responses? And hope, that no other ongoing traffic influences the times in a "non-filterable" way? In my opinion, very optimistic.

So I still think:
Inform the user about the thread and offer alternatives.
But don't "overstate" the thread by only mention a crypto detail.

**Pull request merged**